### ISSUE TITLE
extra-unused-lifetimes should not lint inside macros

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#8895]
### Summary

If we have a macro that generates `impl`s for something, obviously it may look like like the lifetime is unused inside the macro, but it might be used in the macro callsite.

````rust
macro_rules! forward_display_to_print {
    ($($ty:ty),+) => {
        // Some of the $ty arguments may not actually use 'tcx
        $(#[allow(unused_lifetimes)] impl<'tcx> fmt::Display for $ty {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                ty::tls::with(|tcx| {
                    let cx = tcx.lift(*self)
                        .expect("could not lift for printing")
                        .print(FmtPrinter::new(tcx, Namespace::TypeNS))?;
                    f.write_str(&cx.into_buffer())?;
                    Ok(())
                })
            }
        })+
    };
}
````
clippy warns about `<'tcx>` here but removing it breaks the code.

### Lint Name

extra-unused-lifetimes

**Comment by smoelius** [comment]
Are we sure that the proposed behavior is the desired behavior?

Like wouldn't the same reasoning imply that `unused_extern_crates` shouldn't lint inside macros?
```rust
#![warn(unused_extern_crates)]

macro_rules! impl_f {
    ($($stmts:stmt)*) => {
        extern crate rand;
        pub fn f() {
            $($stmts)*
        }
    };
}

mod foo {
    impl_f!(println!("{:?}", rand::random::<u32>()));
}

mod bar {
    impl_f!();
}

fn main() {
    foo::f();
    bar::f();
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=9cf0b1c29ac76d49388ef5767acce101))

**Comment by MrPiao** [comment]
Duplicate of #9014 right?

**Comment by dswij** [comment]
Fixed by #9014. `extra-unused-lifetimes` does not lint the example code anymore in nightly.

Thanks @MrPiao for taking a look

**Comment by oren0e** [comment]
@dswij do you know when this is going to get to stable?

**Comment by dswij** [comment]
@oren0e the fix is in beta. So, we should be getting this in the next stable release
